### PR TITLE
feat(rust): proper parsing for macro args

### LINKED ISSUE #8884: Rust parser does not properly handle token trees within macros
**Describe the bug**
In Rust code, the `matches!($X, ...)` Semgrep pattern does not find instances of `matches!(foo, Foo::A(_))`

**To Reproduce**
Playground example with tests https://semgrep.dev/playground/s/DPrP?editorMode=advanced

**Expected behavior**
The Rust parser properly handles macros and finds and matches all instances.

**What is the priority of the bug to you?**

- [x] P0: blocking your adoption of Semgrep or workflow
- [ ] P1: important to fix or quite annoying
- [ ] P2: regular bug that should get fixed

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR makes it so we can fully parse all expressions which fall into the arguments of Rust function-like macros, or bracketed macros.

## Why:
When we merged https://github.com/returntocorp/semgrep/pull/8209, this was with the acknowledgement that our parsing of Rust macros was "best effort". The way that Rust macros work is that, within the grammar itself, the text is split into "token trees" of the text between the delimiting braces. They are not, however, assembled in any more meaningful way. For instance, `foo!(**x)` does not see `**x`, but instead `*`, `*`, and `x` separately.

This means that putting together these elements is on us. We tried our best to analyze such contents as arguments if we could, but this requires a lot of special casing (such as splitting on `,`, etc) which amounted to approximating a true Rust expression parser.

Thankfully, we already have a Rust expression parser, since we have an entry point into parsing Rust expressions -- via the `__SEMGREP_EXPRESSION` prefix that we insert while parsing patterns. This means that if we can get the text of the macro call's contents, we could feed it back into the parser to try and get something meaningful out of it.

This enables us to properly perform matching and semantic analysis upon Rust programs which make use of macros that have more complicated language constructs within them, specifically those we have not already special cased. An example of where our support for macros was lacking was https://github.com/returntocorp/semgrep/issues/8884#issuecomment-1748723700.

## How:
When parsing a macro such as `foo!(...)`, where the `...` may look like a function call's arguments or not, instead of trying to make sense of the tokens within the tree, we instead will use the ranges to rip out the literal text within the `(...)`, and then prefix it with `__SEMGREP_EXPRESSION` and try to parse it. With luck, this will succeed, if it truly is able to be interpreted as the comma-separated arguments to a function call, which saves us from having to deal with that parsing ourselves.

Some difficulty was induced by the fact that we have to get the ranges right -- all the tokens need to have the same ranges as they appear in the original file, and pattern tokens don't even have byte positions, which makes some things hard. Those also needed to be addressed here.

## Test plan:
`make test`

Closes #8884

### CODE REVIEW

File: languages/rust/generic/Parse_rust_tree_sitter.ml [hunk 1]
```diff
@@ -2107,17 +2147,161 @@ and map_macro_invocation (env : env) ((v1, v2, v3) : CST.macro_invocation) :
...
-  let args =
-    match anys with
-    (* look like a regular function call, just use Arg then *)
-    | [ G.E e ] -> [ G.Arg e ]
-    (* coupling: see `macro_items_to_anys` above *)
-    | [ G.Args args ] -> args
-    | xs -> [ G.OtherArg (("ArgMacro", G.fake ""), xs) ]
-  in
-  G.Call (G.N name |> G.e, (l, args, r)) |> G.e
+  (* Here's where we're going to do something interesting.
```
Comment (aryx) [inline languages/rust/generic/Parse_rust_tree_sitter.ml:2150]: very nice comment

**Review by aryx — APPROVED** [review]
Nice. Smart.
But couldn't we achieve the same thing by extending the tree-sitter-rust grammar
for macro arguments and provide a choice with priority to try an argument list and 
otherwise the token tree thing?
If not can you add a comment why it's not possible?

**Comment by brandonspark** [comment]
@aryx indeed, I should have attempted to fix this upstream via the grammar first. unsure why that didn't occur to me...

but, it will take me a bit before I can get to that, so let's ice this for now.

**Comment by aryx** [comment]
Ok let's merge then

**Pull request merged**